File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
public boolean presentInSubTree(TreeNode root, TreeNode child) {
        if(root == null){
            return false;
        }

        if(root.val == child.val){
            return true;
        }

        return presentInSubTree(root.left, child) || presentInSubTree(root.right, child);
        
    }
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null){
            return null;
        }

        if(root.val == p.val || root.val == q.val){
            return root;
        }

        boolean left = false;
        boolean right = false;

        left = presentInSubTree(root.left, p);
        right =  presentInSubTree(root.right, q);

        if(left != true && right == true){
            return lowestCommonAncestor(root.right, p,q);
        } 

        if(left == true && right != true){
            return lowestCommonAncestor(root.left, p,q);
        }

        return root;

    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        self.count = 0
        def helper(root:TreeNode, val: int):
            if root is None:
                return 
            
            if root.val >= val:
                self.count += 1
                val = max(val,root.val)
            helper(root.left,val)
            helper(root.right,val)
        
        helper(root,-101)
        return self.count
        